### PR TITLE
TypeScript: Convert `pageDataUrls` provider 

### LINKED ISSUE #12668: TypeScript: Convert pageDataUrls provider

### PR DESCRIPTION
**Pull request by timarney**
## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12668

### CODE REVIEW

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx [hunk 1]
```diff
@@ -26,12 +26,8 @@ import useIdleTaskQueue from '../../utils/useIdleTaskQueue';
 import storyPageToDataUrl from '../../utils/storyPageToDataUrl';
 import Context from './context';
 
-/**
- * @typedef {import('../../types.js').Page} Page
- */
-
-function PageDataUrlProvider({ children }) {
-  const [dataUrls, setDataUrls] = useState({});
+function PageDataUrlProvider({ children }: { children: React.ReactNode }) {
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: Let's use the `PropsWithChildren` type from the react package.

File: packages/story-editor/src/utils/storyPageToDataUrl.ts [hunk 2]
```diff
@@ -18,37 +18,38 @@
...
+async function storyPageToDataUrl(
+  page: Page,
+  { width = 400, ...options }: Options
+): Promise<string> {
   const htmlToImage = await import(
     /* webpackChunkName: "chunk-html-to-image" */ 'html-to-image'
   );
 
-  const [node, cleanup] = await storyPageToNode(page, width);
+  const [node, cleanup] = storyPageToNode(page, width);
```
Comment (spacedmonkey) [inline packages/story-editor/src/utils/storyPageToDataUrl.ts]: This doesn't feel right.

Comment (timarney) [reply packages/story-editor/src/utils/storyPageToDataUrl.ts]: Will test was getting flagged with no unnecessary await the top level `storyPageToNode` isn't async

Comment (timarney) [reply packages/story-editor/src/utils/storyPageToDataUrl.ts]: Ah looks like this is because of the current return type
```
* @return {[HTMLElement, Function]} tuple containing DOM node and cleanup method
```

Comment (timarney) [reply packages/story-editor/src/utils/storyPageToDataUrl.ts]: It's now back

https://github.com/GoogleForCreators/web-stories-wp/pull/12727/commits/538d2c898a005bb4a3c3eeacab072df697fcebda

File: packages/story-editor/tsconfig.json [hunk 3]
```diff
@@ -16,6 +16,10 @@
     "src/app/helpCenter/**/*",
     "src/app/history",
     "src/types",
-    "src/utils/usePreventWindowUnload.ts"
+    "src/utils/usePreventWindowUnload.ts",
+    "src/utils/useIdleTaskQueue.ts",
+    "src/utils/idleCallback.ts",
+    "src/utils/storyPageToDataUrl.ts",
```
Comment (miina) [inline packages/story-editor/tsconfig.json]: Using `"src/utils/*.ts",` would make it easier :)

File: packages/story-editor/src/utils/storyPageToDataUrl.ts [hunk 4]
```diff
@@ -48,7 +53,7 @@ async function storyPageToDataUrl(page, { width = 400, ...options }) {
     height: width * (1 / PAGE_RATIO),
     canvasHeight: width * (1 / PAGE_RATIO),
     canvasWidth: width,
-  });
+  } as Options);
```
Comment (miina) [inline packages/story-editor/src/utils/storyPageToDataUrl.ts]: Curious, what's not matching without casting?

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx [hunk 5]
```diff
@@ -41,24 +43,30 @@ function PageDataUrlProvider({ children }) {
...
-      const idleTask = async () => {
-        const dataUrl = await storyPageToDataUrl(storyPage, {});
-        setDataUrls((state) => ({
-          ...state,
-          [storyPage.id]: dataUrl,
-        }));
-      };
+  const queuePageImageGeneration: PageDataUrlsActions['queuePageImageGeneration'] =
+    useCallback(
+      (storyPage: Page) => {
```
Comment (miina) [inline packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: Does it still require adding `Page` here even when using `PageDataUrlsActions['queuePageImageGeneration']` above?

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx [hunk 6]
```diff
@@ -16,22 +16,24 @@
...
-/**
- * @typedef {import('../../types.js').Page} Page
- */
-
-function PageDataUrlProvider({ children }) {
-  const [dataUrls, setDataUrls] = useState({});
+function PageDataUrlProvider({
+  children,
+}: {
+  children: PropsWithChildren<Record<string, never>>;
```
Comment (miina) [inline packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: See [this for context](https://github.com/GoogleForCreators/web-stories-wp/pull/12744#discussion_r1033995957).
```suggestion
  children: PropsWithChildren<unknown>;
```

File: packages/story-editor/src/utils/storyPageToDataUrl.ts (hunk 4 above)
Comment (swissspidy) [reply packages/story-editor/src/utils/storyPageToDataUrl.ts]: Curious as well. Casting hides potential errors.

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx (hunk 5 above)
Comment (timarney) [reply packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: Removed --- your right it wasn't needed.

File: packages/story-editor/src/utils/storyPageToDataUrl.ts (hunk 4 above)
Comment (timarney) [reply packages/story-editor/src/utils/storyPageToDataUrl.ts]: Removed now --- I don't recall the error when I added that but was fixed along the way with some other earlier commits.

File: packages/story-editor/src/utils/storyPageToDataUrl.ts [hunk 7]
```diff
@@ -18,6 +18,8 @@
  * External dependencies
  */
 import { PAGE_RATIO } from '@googleforcreators/units';
+import type { Page } from '@googleforcreators/elements';
+import type { Options } from 'html-to-image/lib/types';
```
Comment (swissspidy) [inline packages/story-editor/src/utils/storyPageToDataUrl.ts]:
```suggestion
import type { Options } from 'html-to-image';
```

This doesn't work?

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx [hunk 8]
```diff
@@ -41,24 +43,30 @@ function PageDataUrlProvider({ children }) {
...
-      return () => {
-        clearQueueOfPageTask();
-      };
-    },
-    [queueIdleTask]
-  );
+        const clearQueueOfPageTask = queueIdleTask([
+          idleTaskUid,
+          idleTask,
+        ]) as () => void;
```
Comment (swissspidy) [inline packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: This casting seems unnecessary. Probably can be resolved by properly typing the whole `useIdleTaskQueue` file

File: packages/story-editor/src/utils/storyPageToDataUrl.ts (hunk 7 above)
Comment (timarney) [reply packages/story-editor/src/utils/storyPageToDataUrl.ts]: I'm seeing this error

Screen Shot 2022-12-01 at 5 33 48 AM

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx (hunk 8 above)
Comment (timarney) [reply packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: Removed -- issue was fixed with a previous commit

File: tests/js/jest.config.js [hunk 9]
```diff
@@ -44,7 +44,7 @@ export default {
   // See https://jestjs.io/docs/configuration#transformignorepatterns-arraystring
   transformIgnorePatterns: ['/node_modules/(?!(use-reduction)/)'],
   testEnvironment: 'jsdom',
-  testMatch: ['**/test/**/*.[jt]s'],
+  testMatch: ['**/test/**/*.[jt]s?(x)'],
```
Comment (miina) [inline tests/js/jest.config.js]: Due to enabling .tsx files, the following tests are now failing:
https://github.com/GoogleForCreators/web-stories-wp/actions/runs/3824643543/jobs/6506964376#step:9:286

Thoughts @barklund @swissspidy?

Since it's not related to this PR, I'll just revert it back, however, these need to get fixes.

Created an issue for it: #12895 12895

File: packages/design-system/tsconfig.json [hunk 10]
```diff
@@ -19,6 +19,7 @@
     "src/types/*",
     "src/types.ts",
     "src/utils/*.ts",
+    "src/utils/*.tsx",
```
Comment (barklund) [inline packages/design-system/tsconfig.json]: Why this change if this PR doesn't actually contain any files from the design system?

File: packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx [hunk 11]
```diff
@@ -16,52 +16,48 @@
...
 import Context from './context';
 
-/**
- * @typedef {import('@googleforcreators/elements').Page} Page
- */
-
-function PageDataUrlProvider({ children }) {
-  const [dataUrls, setDataUrls] = useState({});
+function PageDataUrlProvider({ children }: PropsWithChildren<unknown>) {
+  const [dataUrls, setDataUrls] = useState<PageDataUrlsState['dataUrls']>({});
```
Comment (barklund) [inline packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: Not a big fan of this notation here. I'd rather we just made separate types for these and used those types in the state as well.

Comment (barklund) [reply packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.tsx]: (also because the state only contains this one single prop)

File: packages/story-editor/src/types/pageDataUrls.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+export interface PageDataUrlsActions {
+  queuePageImageGeneration: (Page: Page) => void;
+}
+
+export interface PageDataUrlsContext {
+  state: PageDataUrlsState;
+  actions: PageDataUrlsActions;
+}
```
Comment (barklund) [inline packages/story-editor/src/types/pageDataUrls.ts]: So this would be:

```ts
export type PageDataUrls = Record<string, string>;
export type QueuePageImageGeneration: (page: Page) => void;
export interface PageDataUrlsContext {
  state: { dataUrls: PageDataUrls }
  actions: { queuePageImageGeneration: QueuePageImageGeneration};
}
```

File: packages/story-editor/src/utils/idleCallback.ts [hunk 13]
```diff
@@ -24,7 +24,7 @@ const requestIdleCallback =
     return setTimeout(() => {
       callback({
         get didTimeout() {
-          return options.timeout
+          return typeof options.timeout !== 'undefined'
```
Comment (barklund) [inline packages/story-editor/src/utils/idleCallback.ts:27]: `'timeout' in options` is a bit cleaner.

**Review by barklund — APPROVED** [review]
Looks good, just some comments to make our efforts look more uniform across the codebase.

File: packages/design-system/tsconfig.json (hunk 10 above)
Comment (miina) [reply packages/design-system/tsconfig.json]: Huh, not sure. I took this PR over, perhaps there was a .tsx file originally that was not needed anymore 🤔 

Will remove this.

File: packages/story-editor/src/utils/idleCallback.ts (hunk 13 above)
Comment (miina) [reply packages/story-editor/src/utils/idleCallback.ts:27]: If `options.timeout` is `undefined` but set, `'options' in timeout` would give `true` though.
